src/algebra/module/localized_module.lean

```diff
@@ -0,0 +1,287 @@
+`f` descents to a map `localized_module M S → α`.
+-/
+def lift_on {α : Type*} (x : localized_module S M) (f : M × S → α)
+  (wd : ∀ (p p' : M × S) (h1 : p ≈ p'), f p = f p') : α :=
+quotient.lift_on x f wd
+
+lemma lift_on_mk {α : Type*} {f : M × S → α}
+  (wd : ∀ (p p' : M × S) (h1 : p ≈ p'), f p = f p')
+  (m : M) (s : S) :
+  lift_on (mk m s) f wd = f ⟨m, s⟩ :=
+by convert quotient.lift_on_mk f wd ⟨m, s⟩
+
+/--If `f : M × S → M × S → α` respects the equivalence relation `localized_module.r`, then
+`f` descents to a map `localized_module M S → localized_module M S → α`.
+-/
+def lift_on₂ {α : Type*} (x y : localized_module S M) (f : (M × S) → (M × S) → α)
+  (wd : ∀ (p q p' q' : M × S) (h1 : p ≈ p') (h2 : q ≈ q'), f p q = f p' q') : α :=
+quotient.lift_on₂ x y f wd
+
+lemma lift_on₂_mk {α : Type*} (f : (M × S) → (M × S) → α)
+  (wd : ∀ (p q p' q' : M × S) (h1 : p ≈ p') (h2 : q ≈ q'), f p q = f p' q')
+  (m m' : M) (s s' : S) :
+  lift_on₂ (mk m s) (mk m' s') f wd = f ⟨m, s⟩ ⟨m', s'⟩ :=
+by convert quotient.lift_on₂_mk f wd _ _
+
+instance : has_zero (localized_module S M) := ⟨mk 0 1⟩
+lemma zero_mk (s : S) : mk (0 : M) s = 0 :=
+mk_eq.mpr ⟨1, by rw [one_smul, smul_zero, smul_zero, one_smul]⟩
+
+instance : has_add (localized_module S M) :=
+{ add := λ p1 p2, lift_on₂ p1 p2 (λ x y, mk (y.2 • x.1 + x.2 • y.1) (x.2 * y.2)) $
+    λ ⟨m1, s1⟩ ⟨m2, s2⟩ ⟨m1', s1'⟩ ⟨m2', s2'⟩ ⟨u1, hu1⟩ ⟨u2, hu2⟩, mk_eq.mpr ⟨u1 * u2, begin
+      -- Put everything in the same shape, sorting the terms using `simp`
+      have hu1' := congr_arg ((•) (u2 * s2 * s2')) hu1,
+      have hu2' := congr_arg ((•) (u1 * s1 * s1')) hu2,
+      simp only [smul_add, ← mul_smul, smul_assoc, mul_assoc, mul_comm, mul_left_comm]
+        at ⊢ hu1' hu2',
+      rw [hu1', hu2']
+    end⟩ }
+
+lemma mk_add_mk {m1 m2 : M} {s1 s2 : S} :
+  mk m1 s1 + mk m2 s2 = mk (s2 • m1 + s1 • m2) (s1 * s2) :=
+mk_eq.mpr $ ⟨1, by dsimp only; rw [one_smul]⟩
+
+lemma add_assoc' (x y z : localized_module S M) :
+  x + y + z = x + (y + z) :=
+begin
+  induction x using localized_module.induction_on with mx sx,
+  induction y using localized_module.induction_on with my sy,
+  induction z using localized_module.induction_on with mz sz,
+  simp [mk_add_mk],
+  refine mk_eq.mpr ⟨1, _⟩,
+  rw [one_smul, one_smul],
+  congr' 1,
+  { rw [mul_assoc] },
+  { rw [mul_comm, add_assoc, mul_smul, mul_smul, ←mul_smul sx sz, mul_comm, mul_smul], },
+end
+
+lemma add_comm' (x y : localized_module S M) :
+  x + y = y + x :=
+localized_module.induction_on₂ (λ m m' s s', by rw [mk_add_mk, mk_add_mk, add_comm, mul_comm]) x y
+
+lemma zero_add' (x : localized_module S M) : 0 + x = x :=
+induction_on (λ m s, by rw [← zero_mk s, mk_add_mk, smul_zero, zero_add, mk_eq];
+  exact ⟨1, by rw [one_smul, mul_smul, one_smul]⟩) x
+
+lemma add_zero' (x : localized_module S M) : x + 0 = x :=
+induction_on (λ m s, by rw [← zero_mk s, mk_add_mk, smul_zero, add_zero, mk_eq];
+  exact ⟨1, by rw [one_smul, mul_smul, one_smul]⟩) x
```

Comment:
You may as well make these `private` since no one ever needs them once you have the typeclassses below

Reply:
marked as private